gm4_better_fire/data/better_fire/functions/checks.mcfunction

```diff
@@ -0,0 +1,10 @@
+#check block at flame arrow
+#@s - @e[nbt=!{Fire:-1s},type=arrow]
+#called by better_fire:main
+
+#set fire block if air
+execute if data entity @s inBlockState if block ~ ~ ~ #better_fire:air run playsound minecraft:block.fire.extinguish block @a[distance=..20] ~ ~ ~ 100 2
```

Comment:
You're using this check to play the sound, but instead of only doing that, put the setblock commands from below also under this check, using a function.
But because this is then the only command in this function, you should rather move this command downwards into `main`.